### ISSUE TITLE
Add API to endpoints for connecting a protocol instance

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5270]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#5270|
|Type|enhancement|
|Created|2011-09-21 15:32:52Z|
|Branch|https://github.com/twisted/twisted/tree/connectprotocol-5270-3|

In many cases, having a factory for client connections is unnecessary boilerplate. The endpoint client API should be extended to accept a protocol instance. This would additionally make learning Twisted easier, and allow deprecation of `ClientCreator`.

<details><summary>Searchable metadata</summary>

```
trac-id__5270 5270
type__enhancement enhancement
reporter__itamar itamar
priority__low low
milestone__ 
branch__branches_connectprotocol_5270_3 branches/connectprotocol-5270-3
branch_author__itamarst__tomprince itamarst, tomprince
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1316619172000000 1316619172000000
changetime__1367081699000000 1367081699000000
version__None None
owner__itamar itamar

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_1"></a>|@itamarst set owner to @itamarst<br>@itamarst set status to `assigned`|
|-|-|

Probably better than modifying `connect()`, which wouldn't be backwards compatible, is adding a new API.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

No, adding a new method is just as incompatible.  Watch:

```

>>> from twisted.internet import reactor
>>> from twisted.internet.protocol import Protocol
>>> from twisted.internet.endpoints import clientFromString
>>> clientFromString(reactor, "tcp:localhost:port=1234").connectUsingProtocol(Protocol())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'TCP4ClientEndpoint' object has no attribute 'connectUsingProtocol'
>>> 
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst commented|
|-|-|

I'm not adding a new method, I'm adding a new module-level function, `connectProtocol(endpoint, protocol)`.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

Thanks for clarifying!  FWIW, I think that's the kind of detail that belongs in a ticket description.  Specifications are amazingly useful - particularly when they're _good_ specifications - and good tickets often include good specifications.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst commented|
|-|-|

(In [36495]) Branching to 'connectprotocol-5270'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_6"></a>|@itamarst removed owner<br>@itamarst set status to `new`|
|-|-|

Ready for review, I think.

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/connectprotocol-5270

This provides an implementation of `connectProtocol`, as mentioned above. I updated the client documentation to use the new API instead of requiring a factory, and moved a paragraph about not requiring client factories to the endpoints howto where it seemed more relevant.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @itamarst|
|-|-|

Thanks.  I guess Twisted programmers everywhere will be singing your praises shortly.

  1. `ClientCreator` hasn't been removed yet (not even deprecated), so I disagree with the deletion of the documentation explaining it (I think we've been over this before?).  Documentation should last at least as long as functionality.  People may not be writing code like this anymore, but that doesn't mean they won't need to know how to read it.
  1. Speaking of reading it, how about converting at least one use of `ClientCreator` in Twisted to this new API, to get the ball rolling?  Plus, it's always nice to see that something actually works for some real use case.
  1. arghlhtlje `_endpointspy3.py`... Does any new code really belong here?  I am thinking about how we have a plan to hopefully recover the history of `endpoints.py`, but I don't think it's possible to do that _and_ keep any coherent history for `_endpointspy3.py` (there will only be one file after we fix the situation, after all).  So this inclines me to want all new development to go into `endpoints.py`.  I guess that makes it hard to know whether the new code is Python 3 compatible, and to maintain that compatibility.  But... who cares?
  1. Someone tried to be nice and started an endpoint testing module, `twisted.internet.test.fakeendpoint`.  They didn't get very far, but perhaps it would be nice to expand that rather than introduce single-use, untested fakes in new tests?  Also, I wonder if fake endpoints are even needed in these tests, or if a real endpoint on a fake reactor would work just as well.
  1. I wonder if `connectProtocol` is the best name.  It's not too bad, but part of me really wants to insert a `With` in the middle.  I don't think that actually improves the name, but that doesn't prevent me from being somewhat dissatisfied with the current name.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_8"></a>|@itamarst commented|
|-|-|

(In [36497]) Restore ClientCreator example. Refs [#5270](https://github.com/twisted/twisted/issues/5270)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_9"></a>|@itamarst commented|
|-|-|

(In [36498]) Switch an example to use connectProtocol. Refs [#5270](https://github.com/twisted/twisted/issues/5270)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_10"></a>|@itamarst commented|
|-|-|

(In [37095]) Branching to 'connectprotocol-5270-2'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_11"></a>|@itamarst removed owner|
|-|-|

New build started: https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/connectprotocol-5270-2 - unfortunately looks like some builders are having issues, so this will need to be rerun at some point.

Review points were addressed thusly:

 1. Done.
 1. Done (`ampclient.py`).
 1. No longer relevant.
 1. I switched one test to use `MemoryReactor`. The other test seemed like it'd be more complex using that, and converting `fakeendpoints` to meet my needs would be quite a bit more work.
 1. I left the function as `connectProtocol`, since I was unable to come up with a name that was sufficiently better. Suggestions?

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_12"></a>|@tomprince set owner to @itamarst|
|-|-|

One comment on the documention:
  * If, as the documentation in `doc/core/howto/clients`, `connectProtocol` is the preferred (by which I mean, the way we want to suggest to people that don't know why they might want something else) way to use client endpoints, then it would make sense to explain that first in `doc/core/howto/endpoints`, and then explain the use of factories after, rather than have `connectProtocol` be an aside at the end of the section on clients.
  * [twisted checker error](https://buildbot.twistedmatrix.com/builders/twistedchecker/builds/186/steps/run-twistedchecker/logs/new%20twistedchecker%20errors)
Otherwise, this looks good. Please resubmit after updating the documentation.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_13"></a>|@itamarst commented|
|-|-|

(In [37513]) Address review comments (refs [#5270](https://github.com/twisted/twisted/issues/5270))

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_14"></a>|@itamarst removed owner|
|-|-|

The checker error doesn't strike me as particularly useful; this is test code, not real code, so I am don't think it's worth paying attention to. I did add a docstring, but probably now it will want a rtype or something. Also updated howto.

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/connectprotocol-5270-2

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_15"></a>|@therve commented|
|-|-|

I'm not sure how I feel about that, instead of working on making OneShotFactory public, like what [#5781](https://github.com/twisted/twisted/issues/5781) aims to. Maybe those are not incompatible though, if connectProtocol uses the new public class.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_16"></a>|@tomprince set owner to @itamarst|
|-|-|

Looks good. Merge away.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_17"></a>|@tomprince commented|
|-|-|

(In [38282]) Branching to connectprotocol-5270-3.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_18"></a>|@tomprince set status to `closed`|
|-|-|

(In [38285]) Merge connectprotocol-5270-3: Add API to endpoints for connecting a protocol instance.

Author: itamar
Reviewers: exarkun, tom.prince
Fixes: [#5270](https://github.com/twisted/twisted/issues/5270)

In many cases, having a factory for client connections is
unnecessary boilerplate. The endpoint client API should be
extended to accept a protocol instance. This would
additionally make learning Twisted easier, and allow
deprecation of ClientCreator.
